lib/stud/interval.rb

```diff
@@ -1,30 +1,70 @@
 require "stud/task"
+require "thread"
+
 module Stud
+  STUD_STOP_REQUESTED = :stud_stop_requested
+
+  extend self
+
   # This implementation tries to keep clock more accurately.
   # Prior implementations still permitted skew, where as this one
   # will attempt to correct for skew.
   #
   # The execution patterns of this method should be that
   # the start time of 'block.call' should always be at time T*interval
-  def self.interval(time, opts = {}, &block)
+  def interval(time, opts = {}, &block)
     start = Time.now
     while true
-      break if Task.interrupted?
       if opts[:sleep_then_run]
         start = sleep_for_interval(time, start)
+        break if stop?
         block.call
       else
         block.call
         start = sleep_for_interval(time, start)
+        break if stop?
       end
     end # loop forever
   end # def interval
 
-  def self.sleep_for_interval(time, start)
+  # stop! instructs interval to stop and exit its execution loop before going to
+  # sleep between block executions.
+  # NOW the tricky part is: this is typically an operation that will be called
+  # from another thread than the thread running the interval loop in which case
+  # the target parameter must be set to the Thread object which is running the
+  # interval loop.
+  # Note that the stop logic is compatible with Stud::Task so if interval is run
+  # inside a Stud::Task, calling Stud::Task#stop! will stop the interval the same
+  # way as calling stop! on the interval itself.
+  # @param [Thread] the target thread to stop, defaut to Thread.current
+  def stop!(target = Thread.current)
+    target[STUD_STOP_REQUESTED] = true
+    target.wakeup
+  end
+
+  # stop? returns true if stop! has been called
+  # @param [Thread] the target thread to check for stop, defaut to Thread.current
+  # @return [Boolean] true if the stop! has been called
+  def stop?(target = Thread.current)
+    target[STUD_STOP_REQUESTED]
+  end
+  alias_method :interrupted?, :stop?
+
+  private
+
+  def sleep_for_interval(time, start)
     duration = Time.now - start
     # Sleep only if the duration was less than the time interval
     if duration < time
-      sleep(time - duration)
+      seconds, reminder = (time - duration).divmod 1
+
+      # to avoid race conditions, check for stop every second
+      seconds.times.each do |i|
```

Comment:
@jsvd for your idea of a state variable to signal when entering a sleep, you will still face potential race conditions and will have to introduce synchronization constructs to try to prevent such race condition and you will face the fact that because `Stud.interval` is "stateless" (not a proper class etc) it is non-trivial to introduce a mutex that can be shared between the caller thread and the interval running thread etc. So unless we refactor completely (which I tried to avoid here) I don't think there are other pragmatic options.